### ISSUE TITLE
How to get the coordinates of all 3D Gaussian points on the plane after each iteration

### ISSUE THREAD

**Issue by 520jz** [graphdeco-inria/gaussian-splatting#698]
Hello, I checked the code and found that there is an all-zero matrix screen_point(means2D), which is used to store the gradient of the two-dimensional Gaussian ellipsoid, but when propagating forward, I found that this scree_point(means2D) is not passed into CUDA, which causes the array in it to be an all-0 matrix all the time, and can only get its gradient but not the coordinates
![image](https://github.com/graphdeco-inria/gaussian-splatting/assets/125104736/a76fc4e7-b141-44a1-9500-ab2ecb5cc949)
Now I want to get the coordinates of all Gaussian ellipsoids on a 2D plane, is there any way to do it？

**Comment by JerryPW** [comment]
Hi! I have the same question with you. Have you figured it out?

**Comment by shreyesss** [comment]
Same question here? Please let me know if you've figured this out? 
I am trying to visualize all the gaussian splats in 2D projected onto my image plane

**Comment by harveyyang66** [comment]
Hi, I have the same question, may I ask if the problem is sorted already?

**Comment by fengaba1818** [comment]
I have the same question.

**Comment by DI-LEE** [comment]
I have exactly same question..!

**Comment by Martisum** [comment]
@520jz I have a way, but I haven't tried it yet. I plan to try it in the near future. If it succeeds, I will continue to add it below.
During the debugging process, I noticed that even if "means2D" is empty, we can still manually calculate the value of "means2D". The specific steps are as follows:
1. **The length of "radii" and "visibility_filter" is consistent with "gaussians.get_xyz", that is, the subscript of the "radii" sequence corresponds to the subscript of the center coordinate of the Gaussian stored in the Gaussian model.** Through "visibility_filter", we can know which Gaussians are rendered in the view cone, and the information of this Gaussian can be obtained through the "gaussians" object.
2. Because a random camera is selected for each rendering, you can find the specific information of the corresponding rotation matrix R and offset t in the "viewpoint_cam" object, so that you can transform the Gaussian ellipsoid to the camera coordinate system.
3. The FoVx ​​and FoVy information are also recorded in the "viewpoint_cam" object, and you can use this to convert the information of the camera coordinate system into pixel information. The conversion can be completed using the following formula:

- tanfovx = math.tan(math.radians(viewpoint_camera.FoVx) * 0.5)
- screen_x = X / (Z * tanfovx)

At this point, you have obtained the coordinates of this Gaussian body on the two-dimensional pixel plane

**Comment by 2mha0** [comment]
@Martisum 
Thank you for your suggestion. I have successfully implemented the logic in train.py. The code snippet is as follows:
```python
        render_pkg = render(viewpoint_cam, gaussians, pipe, bg, use_trained_exp=dataset.train_test_exp, separate_sh=SPARSE_ADAM_AVAILABLE)
        image, viewspace_point_tensor, visibility_filter, radii = render_pkg["render"], render_pkg["viewspace_points"], render_pkg["visibility_filter"], render_pkg["radii"]
        depth, alpha, norm, extra = render_pkg["depth"], render_pkg["alpha"], render_pkg["norm"], render_pkg["extra"]

        _xyz_world = gaussians.get_xyz
        # 筛选出可见的高斯点 (仍是世界坐标)
        visible_xyz_world = _xyz_world[visibility_filter.squeeze()]

        R_cw = torch.from_numpy(viewpoint_cam.R).to(visible_xyz_world.device).float()
        t_cw = viewpoint_cam.camera_center

        points_translated_to_camera_origin = visible_xyz_world - t_cw.unsqueeze(0)

        visible_xyz_cam = torch.matmul(points_translated_to_camera_origin, R_cw)

        # 获取可见点的X、Y、Z坐标
        X = visible_xyz_cam[:, 0]  # 所有可见点的X坐标
        Y = visible_xyz_cam[:, 1]  # 所有可见点的Y坐标
        Z = visible_xyz_cam[:, 2]  # 所有可见点的Z坐标

        tanfovx = math.tan(viewpoint_cam.FoVx * 0.5)
        tanfovy = math.tan(viewpoint_cam.FoVy * 0.5)

        # 使用公式计算屏幕空间坐标
        screen_x = X / (Z * tanfovx)
        screen_y = Y / (Z * tanfovy)

        cx = viewpoint_cam.image_width // 2
        cy = viewpoint_cam.image_height // 2
        pixel_x = screen_x * (viewpoint_cam.image_width / 2) + cx
        pixel_y = screen_y * (viewpoint_cam.image_height / 2) + cy
        pixel_points = torch.stack([pixel_x, pixel_y], dim=1) 
```
Then I verified the correctness of the code by plotting the transformed point coordinates on the rendered image.

<img width="1600" height="1199" alt="Image" src="https://github.com/user-attachments/assets/365da5db-ccdf-47af-944e-b58cbb5c862d" />
<img width="1600" height="1199" alt="Image" src="https://github.com/user-attachments/assets/8a3c4aaf-1e21-4050-ab0c-679616b902fd" />
It can be seen that the points have been successfully mapped onto images with different poses, and as the number of Gaussian points increases with the densification strategy, the number of points indeed increases.

**Comment by XiShuFan** [comment]
thanks @2mha0 your script works
